contrib/ci/oss-fuzz.py

```diff
@@ -0,0 +1,431 @@
+            "-Dinternal_pcre=true",
+        ],
+    )
+    builder.add_work_includedir("include/glib-2.0")
+    builder.add_work_includedir("lib/glib-2.0/include")
+    builder.add_build_ldflag("lib/libgio-2.0.a")
+    builder.add_build_ldflag("lib/libgmodule-2.0.a")
+    builder.add_build_ldflag("lib/libgobject-2.0.a")
+    builder.add_build_ldflag("lib/libglib-2.0.a")
+    builder.add_build_ldflag("lib/libgthread-2.0.a")
+
+    # JSON-GLib
+    src = builder.checkout_source(
+        "json-glib",
+        url="https://gitlab.gnome.org/GNOME/json-glib.git",
+        commit="c30c988ac3b0d9e7c332232e3c3969818749b239",
+    )
+    builder.build_meson_project(
+        src,
+        [
+            "-Dgtk_doc=disabled",
+            "-Dtests=false",
+            "-Dintrospection=disabled",
+        ],
+    )
+    builder.add_work_includedir("include/json-glib-1.0/json-glib")
+    builder.add_work_includedir("include/json-glib-1.0")
+    builder.add_build_ldflag("lib/libjson-glib-1.0.a")
+
+    # libxmlb
+    src = builder.checkout_source(
+        "libxmlb", url="https://github.com/hughsie/libxmlb.git"
+    )
+    builder.build_meson_project(
+        src,
+        [
+            "-Dgtkdoc=false",
+            "-Dintrospection=false",
+            "-Dtests=false",
+        ],
+    )
+    builder.add_work_includedir("include/libxmlb-2")
+    builder.add_work_includedir("include/libxmlb-2/libxmlb")
+    builder.add_build_ldflag("lib/libxmlb.a")
+
+    # config.h
+    builder.write_header(
+        "config.h",
+        {
+            "FWUPD_DATADIR": "/tmp",
+            "FWUPD_LOCALSTATEDIR": "/tmp",
+            "FWUPD_PLUGINDIR": "/tmp",
+            "FWUPD_SYSCONFDIR": "/tmp",
+            "HAVE_REALPATH": None,
+            "HAVE_UTSNAME_H": None,
+            "PACKAGE_NAME": "fwupd",
+            "PACKAGE_VERSION": "0.0.0",
+        },
+    )
+
+    # version
+    builder.write_header(
+        "libfwupd/fwupd-version.h",
+        {
+            "FWUPD_MAJOR_VERSION": 0,
+            "FWUPD_MINOR_VERSION": 0,
+            "FWUPD_MICRO_VERSION": 0,
+            "FWUPD_CHECK_VERSION(major,minor,micro)": 1,
+        },
+    )
+
+    # libfwupd
+    builder.add_src_includedir("fwupd")
+    builder.add_src_includedir("fwupd/libfwupd")
+    builder.add_src_includedir("fwupd/libfwupdplugin")
+    for obj in [
+        "fwupd-common",
+        "fwupd-device",
+        "fwupd-enums",
+        "fwupd-error",
+        "fwupd-release",
+    ]:
+        builder.compile("fwupd/libfwupd/{}.c".format(obj), obj)
+
+    # libfwupdplugin
+    for obj in [
+        "fu-chunk",
+        "fu-common",
+        "fu-common-version",
+        "fu-device",
+        "fu-device-locker",
+        "fu-firmware",
+        "fu-firmware-common",
+        "fu-firmware-image",
+        "fu-dfu-firmware",
+        "fu-fmap-firmware",
+        "fu-dfuse-firmware",
+        "fu-srec-firmware",
+        "fu-ihex-firmware",
+        "fu-quirks",
+        "fu-volume",
+    ]:
```

Comment:
If there is going to be a different build system for oss-fuzz, can it at least browse directories to find the source files it compiles rather than hardcoding in _another_ place?

Reply:
I do wonder if we can use `meson introspect` -- I do agree it sucks to have the list in two places.

Reply:
I had a play with this; `meson introspect meson.build --projectinfo` gives us the project version (which we don't actually need, as FWUPD_CHECK_VERSION always returns `1`), but even the AST output doesn't give you the variables like `libfwupd_src`. The important bit is we don't want to build everything like the meson files does; only maybe 30-50% of the sources. I guess there are two things we could do here:

 * We have something like this in the meson.build:

	+++ libfwupd/meson.build
	@@ -40,12 +40,12 @@ endif
	 libfwupd_src = [
	   'fwupd-client.c',
	   'fwupd-client-sync.c',
	-  'fwupd-common.c',
	+  'fwupd-common.c',         # used for fuzzers
	   'fwupd-device.c',
	-  'fwupd-enums.c',
	-  'fwupd-error.c',
	+  'fwupd-enums.c',          # used for fuzzers
	+  'fwupd-error.c',          # used for fuzzers
	   'fwupd-security-attr.c',
	-  'fwupd-release.c',
	+  'fwupd-release.c',        # used for fuzzers
	   'fwupd-plugin.c',
	   'fwupd-remote.c',
	 ]

..and then just read the meson.build file and create the file list based on the source lines with the magic comment suffix.

* We source a file like libfwupd/fuzzing.py and libfwupdplugin/fuzzing.py to get the filelists -- although this isn't awesome as you still need to remember to edit the file list in two places.

Reply:
> with the magic comment

That actually worked quite well. I've pushed a new version that does that.